javascript/ql/src/semmle/javascript/security/performance/SuperlinearBackTracking.qll

```diff
@@ -0,0 +1,141 @@
+ * perform backtracking in superlinear time.
+ */
+
+import javascript
+
+/**
+ * A regular expression term that permits unlimited repetitions.
+ */
+private class InfiniteRepetitionQuantifier extends RegExpQuantifier {
+  InfiniteRepetitionQuantifier() {
+    this instanceof RegExpPlus
+    or
+    this instanceof RegExpStar
+    or
+    this instanceof RegExpRange and not exists(this.(RegExpRange).getUpperBound())
+  }
+}
+
+/**
+ * Holds if `t` matches at least an epsilon symbol.
+ *
+ * That is, this term does not restrict the language of the enclosing regular expression.
+ *
+ * This is implemented as an under-approximation, and this predicate does not hold for sub-patterns in particular.
+ */
+private predicate matchesEpsilon(RegExpTerm t) {
+  t instanceof RegExpStar
+  or
+  t instanceof RegExpOpt
+  or
+  t.(RegExpRange).getLowerBound() = 0
+  or
+  exists(RegExpTerm child |
+    child = t.getAChild() and
+    matchesEpsilon(child)
+  |
+    t instanceof RegExpAlt or
+    t instanceof RegExpGroup or
+    t instanceof RegExpPlus or
+    t instanceof RegExpRange
+  )
+  or
+  matchesEpsilon(t.(RegExpBackRef).getGroup())
+  or
+  forex(RegExpTerm child | child = t.(RegExpSequence).getAChild() | matchesEpsilon(child))
+}
+
+/**
+ * Gets a term that matches the symbol immediately before `t` is done matching.
+ */
+private RegExpTerm getAMatchPredecessor(RegExpTerm t) {
+  result = t
+  or
+  exists(RegExpTerm recurse | result = getAMatchPredecessor(recurse) |
+    // wrappers depend on their children
+    recurse = t.getAChild() and
+    (
+      t instanceof RegExpAlt
+      or
+      t instanceof RegExpGroup
+      or
+      t instanceof RegExpQuantifier
+    )
+    or
+    recurse = t.(RegExpSequence).getLastChild()
+    or
+    recurse = t.(RegExpBackRef).getGroup()
+    or
+    // recurse past epsilon terms
+    matchesEpsilon(t) and recurse = t.getPredecessor()
+  )
+}
+
+private RegExpCharacterClassEscape unwrapCharacterClassEscape(RegExpTerm t) {
+  t = result or
+  t.(RegExpCharacterClass).getAChild() = result
+}
+
+pragma[inline]
+private predicate compatibleConstants(RegExpTerm t1, RegExpTerm t2) {
+  exists(string s1, string s2 |
+    s1 = t1.getAMatchedString() and s2 = t2.getAMatchedString()
+    or
+    unwrapCharacterClassEscape(t1).getValue() = s1 and
+    unwrapCharacterClassEscape(t2).getValue() = s2
+  |
+    s1 = s2
+  )
+}
+
+/**
+ * Holds if `s1` and `s2` possibly have a non-empty intersection.
+ *
+ * This is a simple, and under-approximate, version of
+ * ReDoS::compatible/2, as this predicate only handles some character
+ * classes and constant values.
+ */
+pragma[inline]
+private predicate compatible(RegExpTerm s1, RegExpTerm s2) {
+  not s1.(RegExpCharacterClass).isInverted() and
+  not s2.(RegExpCharacterClass).isInverted() and
+  compatibleConstants(s1, s2)
+  or
+  s1.(RegExpCharacterClass).isInverted() and
+  s2.(RegExpCharacterClass).isInverted() and
+  compatibleConstants(s1.getAChild(), s2.getAChild())
```

Comment:
Did you intend not to handle non-inverted char classes at all?

The check isn't technically correct for inverted character classes, but would be correct for non-inverted ones, so I'm curious if you meant `isNonInverted()` instead of `isInverted()`. (`isNonInverted()` doesn't exist but could be added)

Reply:
Huh. I thought I had eliminated all inversion handling.. Let me remove this disjunct..

> Did you intend not to handle non-inverted char classes at all?

I handle the non-inverted char clasess just above this disjunct: `not s1.(RegExpCharacterClass).isInverted()` ...